services/src/main/java/io/scalecube/services/RemoteServiceInstance.java

```diff
@@ -0,0 +1,163 @@
+package io.scalecube.services;
+
+import io.scalecube.cluster.ICluster;
+import io.scalecube.transport.Address;
+import io.scalecube.transport.Message;
+
+import com.google.common.base.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+public class RemoteServiceInstance implements ServiceInstance {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RemoteServiceInstance.class);
+
+  private final ICluster cluster;
+  private final Address address;
+  private final String memberId;
+  private final String serviceName;
+
+  /**
+   * Remote service instance constructor to initiate instance.
+   * 
+   * @param cluster to be used for instance context.
+   * @param serviceReference service reference of this instance.
+   */
+  public RemoteServiceInstance(ICluster cluster, ServiceReference serviceReference) {
+    this.serviceName = serviceReference.serviceName();
+    this.cluster = cluster;
+    this.address = serviceReference.address();
+    this.memberId = serviceReference.memberId();
+  }
+
+  @Override
+  public String serviceName() {
+    return serviceName;
+  }
+
+  @Override
+  public Object invoke(Message request, ServiceDefinition definition) throws Exception {
+    Preconditions.checkArgument(definition != null, "Service definition can't be null");
+
+    // Try to call via messaging
+    // Request message
+    if (definition.returnType().equals(CompletableFuture.class)) {
+      if (definition.parametrizedType().equals(Message.class)) {
+        return futureInvokeMessage(request);
+      } else {
+        return futureInvokeGeneric(request);
```

Comment:
Correct me if I wrong, but looks like these two invocations _futureInvokeMessage_  and _futureInvokeGeneric_ are pretty the same, the only differenece is what should be result of completable future message or message.data. I suggest to avoid such duplication of code, and use some generic method with message transformer function in parameters. e.g futureInvoke(request, message -> message.data())